include/lbann/weights/weights_proxy.hpp

```diff
@@ -0,0 +1,170 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+#ifndef LBANN_WEIGHTS_WEIGHTS_PROXY_HPP_INCLUDED
+#define LBANN_WEIGHTS_WEIGHTS_PROXY_HPP_INCLUDED
+
+#include "lbann/weights/weights.hpp"
+
+namespace lbann {
+
+/** @class WeightsProxy
+ *  @brief Proxy a weights object as a different data type.
+ *
+ *  Access is read-only; updates to the proxied weights must be
+ *  handled directly through the origin object.
+ *
+ *  I'm thinking about ways to enforce this without users having to
+ *  think (how horrible would that be!), but right now the contract
+ *  for this class is that the weights are read from the master copy
+ *  at the beginning of forward prop and then remain constant
+ *  throughout the iteration. At the end of the iteration, the proxy
+ *  values are considered invalid once gradient updates have begun.
+ *
+ *  @tparam TensorDataType The type to which the weights are proxied.
+ */
+template <typename TensorDataType>
+class WeightsProxy
+{
+  using DataTypeWeights = data_type_weights<TensorDataType>;
+  using ValuesType = El::AbstractDistMatrix<TensorDataType>;
```

Comment:
Outside the scope of this PR: It would be nice to add template arguments so `ValuesType` is a concrete `DistMatrix`. This could handle cases where CPU and GPU layers can share the same weights, or two layers with different data layouts. We would need to update all the layers to manually specify the data layout they need (unlike our current approach, where they just hope it already is the way they need it).

Reply:
Given the current design of the larger LBANN, I'm not sure this would be valuable without a lot of extra effort. Even though matrix distributions are statically determinable (they appear to be functions of `data_layout`), we treat them dynamically. So unless we make this polymorphic with the construction similarly (dynamically) branched, I think this would be hard. A temporary alternative might be to take a (dynamically-determined) `DistData` object that could be factored into the "view or copy" decision.